pkg/storage/stores/shipper/compactor/deletion/delete_requests_client.go

```diff
@@ -0,0 +1,174 @@
+
+func WithRequestClientCacheDuration(d time.Duration) DeleteRequestsStoreOption {
+	return func(c *deleteRequestsClient) {
+		c.cacheDuration = d
+	}
+}
+
+func NewDeleteRequestsClient(addr string, c doer, opts ...DeleteRequestsStoreOption) (DeleteRequestsClient, error) {
+	u, err := url.Parse(addr)
+	if err != nil {
+		level.Error(log.Logger).Log("msg", "error parsing url", "err", err)
+		return nil, err
+	}
+	u.Path = getDeletePath
+
+	client := &deleteRequestsClient{
+		url:           u.String(),
+		httpClient:    c,
+		cacheDuration: 5 * time.Minute,
+		cache:         make(map[string][]DeleteRequest),
+		stopChan:      make(chan struct{}),
+	}
+
+	for _, o := range opts {
+		o(client)
+	}
+
+	go client.updateLoop()
+	return client, nil
+}
+
+func (c *deleteRequestsClient) GetAllDeleteRequestsForUser(ctx context.Context, userID string) ([]DeleteRequest, error) {
+	if cachedRequests, ok := c.getCachedRequests(userID); ok {
+		return cachedRequests, nil
+	}
+
+	requests, err := c.getRequestsFromServer(ctx, userID)
+	if err != nil {
+		return nil, err
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.cache[userID] = requests
+
+	return requests, nil
+}
+
+func (c *deleteRequestsClient) getCachedRequests(userID string) ([]DeleteRequest, bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	res, ok := c.cache[userID]
+	return res, ok
+}
+
+func (c *deleteRequestsClient) Stop() {
+	close(c.stopChan)
+}
+
+func (c *deleteRequestsClient) updateLoop() {
+	t := time.NewTicker(c.cacheDuration)
+	for {
+		select {
+		case <-t.C:
+			c.updateCache()
+		case <-c.stopChan:
+			return
+		}
+	}
+}
+
+func (c *deleteRequestsClient) updateCache() {
+	userIDs := c.currentUserIDs()
+
+	newCache := make(map[string][]DeleteRequest)
+	for _, userID := range userIDs {
+		deleteReq, err := c.getRequestsFromServer(context.Background(), userID)
+		if err != nil {
+			level.Error(log.Logger).Log("msg", "error getting delete requests from the store", "err", err)
+			continue
+		}
+		newCache[userID] = deleteReq
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.cache = newCache
+}
+
+func (c *deleteRequestsClient) currentUserIDs() []string {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	userIDs := make([]string, 0, len(c.cache))
+	for userID := range c.cache {
+		userIDs = append(userIDs, userID)
+	}
+
+	return userIDs
+}
+
+func (c *deleteRequestsClient) getRequestsFromServer(ctx context.Context, userID string) ([]DeleteRequest, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.url, nil)
+	if err != nil {
+		level.Error(log.Logger).Log("msg", "error getting delete requests from the store", "err", err)
+		return nil, err
+	}
+
+	req.Header.Set(orgHeaderKey, userID)
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		level.Error(log.Logger).Log("msg", "error getting delete requests from the store", "err", err)
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
```

Comment:
This can be written like this as well:

```suggestion
	if resp.StatusCode/100 != 2 {
```

Reply:
Nice, committed